### PR TITLE
Add mortar based EqualGradient constraint

### LINKED ISSUE #6807: Allow weak enforcement of periodicity
### Description of the enhancement or error report

For mechanics problems I need a boundary condition that (weakly) enforces the periodicity of the strain (i.e. the _gradients_ of the displacement variables). I was thinking to create a new class of integrated boundary conditions that has access to both faces of the elements touching a periodic boundary on either side. This would be somewhat similar to the `FaceFaceConstraint`s in the mortar system.

### Rationale for the enhancement or information for reproducing the error

Such a BC will enable further degrees of freedom (in particular will it allow deformation that is not volume conserving).

### Identified impact

New capability.

### PR DESCRIPTION
**Pull request by dschwen**
Enforce gradient equality across a mortar interface. This is a new type of periodic boundary condition that allows us to set up strain periodic representative volume elements with commensurate interfaces while supporting volume changes. Other applications may be supporting periodicity in electrical fields defined by their underlying scalar potential.
- adds the `EqualGradeintConstraint`
- adds support for `_grad_u_master`, `_grad_u_slave`, `_grad_test_master`, `_grad_test_slave` to `FaceFaceConstraint`
- adds `_side_grad_phi` to `PenetrationInfo`
- adds `MooseVariable::getGradient`
- adds `MortarPeriodicMesh`, which currently is a simplified version of `GeneratedMesh` (only works in 2D with first order elements quad/edge) that adds two line meshes for the mortar interfaces used in the periodicity constraints.

eigenstrain_elasticenergy

I plan to spend more time on `MortarPeriodicMesh` (ensure feature parity with `GeneratedMesh` by adding the ability to  combine multiple meshes and use the libMesh routines to generate the interfaces). Right now it is bare bones to facilitate testing without having to fire up Cubit for every resolution change.

diffusion

Thanks to @jiangwen84, @cpritam, and @snschune for helpful discussions!

Closes #6807

### CODE REVIEW

File: framework/include/constraints/EqualGradientConstraint.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+#include "FaceFaceConstraint.h"
+
+class EqualGradientConstraint;
+
+template<>
+InputParameters validParams<EqualGradientConstraint>();
+
+/**
+ *
```
Comment (dschwen) [inline framework/include/constraints/EqualGradientConstraint.h]: oops...

**Comment by permcody** [comment]
This is pretty cool. Also we have "Mortar" work written into our LDRD so I really want to support this work. If there's anything you need from the MOOSE team, let us know.

**Comment by dschwen** [comment]
The first thing that comes to mind would be more error checking (and user friendly error messages for the mortar system). If you define an interface and put nor variables on it you get a PETSc error. If you put the variable on the wrong interface you get a segfault (if I recall correctly).

**Comment by permcody** [comment]
Let's capture these in tickets. I'll add it to my own list of things to look at. It's either David or Me that will work on this stuff.

File: modules/phase_field/src/mesh/MortarPeriodicMesh.C [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    _nx(other_mesh._nx),
+    _ny(other_mesh._ny),
+    _nz(other_mesh._nz),
+    _xmin(getParam<Real>("xmin")),
+    _xmax(getParam<Real>("xmax")),
+    _ymin(getParam<Real>("ymin")),
+    _ymax(getParam<Real>("ymax")),
+    _zmin(getParam<Real>("zmin")),
+    _zmax(getParam<Real>("zmax")),
+    _us_mesh(dynamic_cast<UnstructuredMesh&>(getMesh()))
```
Comment (andrsd) [inline modules/phase_field/src/mesh/MortarPeriodicMesh.C]: ` ` before `&`

File: modules/phase_field/src/mesh/MortarPeriodicMesh.C [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+#include "MortarPeriodicMesh.h"
+// #include "NonlinearSystem.h"
+
+// libMesh includes
+// #include "libmesh/getpot.h"
+#include "libmesh/mesh_generation.h"
+#include "libmesh/string_to_enum.h"
+#include "libmesh/edge_edge2.h"
+// #include "libmesh/periodic_boundaries.h"
+// #include "libmesh/periodic_boundary_base.h"
```
Comment (andrsd) [inline modules/phase_field/src/mesh/MortarPeriodicMesh.C]: Remove the commented out `#include`s

File: modules/phase_field/src/mesh/MortarPeriodicMesh.C [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+{
+  // hard code teh elem type for now (otherwise I have to duplicate too much stuff from libmesh)
+  ElemType elem_type_2D = Utility::string_to_enum<ElemType>("QUAD4");
+  ElemType elem_type_3D = Utility::string_to_enum<ElemType>("HEX8");
+
+  switch (_dim)
+  {
+    case 2:
+      // build the main mesh
+      MeshTools::Generation::build_square(dynamic_cast<UnstructuredMesh&>(getMesh()),
```
Comment (andrsd) [inline modules/phase_field/src/mesh/MortarPeriodicMesh.C]: A space in front of `&`.

File: framework/src/geomsearch/PenetrationInfo.C [hunk 5]
```diff
@@ -23,8 +23,16 @@
 #include "Moose.h"
 #include "MooseMesh.h"
 
-PenetrationInfo::PenetrationInfo(const Node * node, const Elem * elem, Elem * side, unsigned int side_num, RealVectorValue norm, Real norm_distance, Real tangential_distance, const Point & closest_point, const Point & closest_point_ref, const Point & closest_point_on_face_ref, std::vector<Node*> off_edge_nodes, const std::vector<std::vector<Real> > & side_phi, const std::vector<RealGradient> & dxyzdxi, const std::vector<RealGradient> & dxyzdeta, const std::vector<RealGradient> & d2xyzdxideta)
-  : _node(node),
+PenetrationInfo::PenetrationInfo(const Node * node, const Elem * elem, Elem * side, unsigned int side_num,
+                                 RealVectorValue norm, Real norm_distance, Real tangential_distance,
+                                 const Point & closest_point, const Point & closest_point_ref, const Point & closest_point_on_face_ref,
+                                 std::vector<Node*> off_edge_nodes,
```
Comment (andrsd) [inline framework/src/geomsearch/PenetrationInfo.C]: A space before `*`

File: modules/combined/examples/mortar/eigenstrain.i [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+
+  #petsc_options_iname = '-pc_type -sub_pc_type '
+  #petsc_options_value = 'asm      ilu'
+
+  line_search = basic
+  petsc_options_iname = '-pc_type -pc_factor_shift_type -pc_factor_shift_amount'
+  petsc_options_value = ' lu       NONZERO               1e-10'
+
+  #petsc_options_iname = '-ksp_type -pc_type -sub_pc_type -sub_pc_factor_shift_type -sub_pc_factor_shift_amount -pc_asm_overlap'
+  #petsc_options_value = 'gmres      asm      lu           NONZERO                  1e-10                       1             '
```
Comment (andrsd) [inline modules/combined/examples/mortar/eigenstrain.i]: Are the commented out PETSc options good for something?

File: modules/combined/examples/mortar/eigenstrain.i [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+  exodus = true
+  [./table]
+    type = CSV
+    delimiter = ' '
+  [../]
+[]
+
+[Debug]
+  # show_var_residual_norms = true
+[]
```
Comment (andrsd) [inline modules/combined/examples/mortar/eigenstrain.i]: This can be removed

**Comment by jwpeterson** [comment]
> Thanks to @jiangwen84, @cpritam, and @snschune for helpful discussions!

You forgot to @ me, bro.

**Comment by dschwen** [comment]
You were no help whatsoever

**Comment by dschwen** [comment]
and @god

**Comment by god** [comment]
This project has been blessed.

**Comment by permcody** [comment]
👍

**Comment by mangerij** [comment]
I looked at the tests for this (I'm of the opinion that we'll need to "try" this capability pretty soon). Can this be used on a 3D mesh with quasi-periodic boundary conditions?

It seems your 3D test only enforces this on a small 2D subdomain embedded in between two 3D meshes.

For example, in the language of GeneratedMesh, I'd like to have periodicity of the total strains(inhomogeneous du_i/dx_j and eigenstrain ) across the boundaries 'top bottom left right' but enforce no such condition on 'front back' ?

Do I just set subdomain  = '0'  and then master = left, slave = 'right' ? and then another for top, bottom? How do I restrict this to only be the BC on displacements and not other order parameters in my input file?

**Comment by dschwen** [comment]
It can be used on 3D meshes. The `MortarPeriodicMesh` only implements the 2D case so far (i should update this). The whole mortar business is a bit more complex. You tie together two interfaces in the simulation. To do that you need an additional lower dimensional mesh that matches the two interfaces and on which a lagrange multiplier variable lives. You need lagrange multiplier for each variable (or gradient component) that you want to restrict. You cannot accidentally apply this to too many variables.

**Comment by mangerij** [comment]
I understand, I would use a total of four lower dimensional meshes on the 'top bottom left right' boundaries.

And 'lm' in your test files is your lagrange multiplier; so I assume I'd need 3 for disp_x disp_y disp_z.

I'll bring this up at our next meeting. 

My only remaining question is, since in your tests you have a lower dimensional mesh wedged between two other meshes explicitly, is this possible to even do where 'top' and 'bottom' are the faces that this low dim mesh shares with the 3D mesh? Again, talking about the simple GeneratedMesh here.

**Comment by dschwen** [comment]
You need _one_ additional mesh per _pair_of interfaces that you tie together. And _each_ mesh needs a lagrange multiplier for each variable (but those variables are restricted to the mortar meshes so this is not adding more DOFs than having the same LM on all meshes).

**Pull request merged**